--- tcl125.hal	2021-05-31 16:46:45.461253482 +0100
+++ /root/tcl125.hal.with-spindle	2021-05-31 16:46:40.501327874 +0100
@@ -6,7 +6,7 @@
 loadrt [KINS]KINEMATICS
 loadrt [EMCMOT]EMCMOT servo_period_nsec=[EMCMOT]SERVO_PERIOD num_joints=[KINS]JOINTS
 loadrt hostmot2
-loadrt hm2_eth board_ip="10.10.10.10" config=" num_encoders=1 num_pwmgens=0 num_stepgens=4 sserial_port_0=00xxxx" 
+loadrt hm2_eth board_ip="10.10.10.10" config=" num_encoders=1 num_pwmgens=1 num_stepgens=4 sserial_port_0=00xxxx" 
 setp    hm2_7i96.0.watchdog.timeout_ns 5000000
 loadrt pid names=pid.x,pid.z,pid.s
 
@@ -133,6 +133,19 @@
 net z-neg-limit     =>  joint.1.neg-lim-sw-in
 net z-pos-limit     =>  joint.1.pos-lim-sw-in
 
+# Stepper 2 - toolchanger
+
+setp   hm2_7i96.0.stepgen.02.dirhold         [JOINT_1]DIRHOLD
+setp   hm2_7i96.0.stepgen.02.steplen         [JOINT_1]STEPLEN
+setp   hm2_7i96.0.stepgen.02.stepspace       [JOINT_1]STEPSPACE
+setp   hm2_7i96.0.stepgen.02.position-scale  530
+setp   hm2_7i96.0.stepgen.02.step_type        0
+setp   hm2_7i96.0.stepgen.02.control-type     0
+setp   hm2_7i96.0.stepgen.02.maxaccel         [JOINT_1]STEPGEN_MAXACCEL
+setp   hm2_7i96.0.stepgen.02.maxvel           [JOINT_1]STEPGEN_MAXVEL
+setp   hm2_7i96.0.stepgen.02.enable true
+
+
 #*******************
 #  SPINDLE
 #*******************
@@ -148,8 +161,12 @@
 setp   pid.s.maxoutput [SPINDLE_0]MAX_OUTPUT
 setp   pid.s.error-previous-target true
 
+
+setp hm2_7i96.0.pwmgen.00.value 0.1
+setp hm2_7i96.0.pwmgen.pwm_frequency 2000
+
 net spindle-index-enable  <=> pid.s.index-enable
-net spindle-enable        =>  pid.s.enable
+net spindle-enable        =>  hm2_7i96.0.pwmgen.00.enable
 net spindle-vel-cmd-rpm     => pid.s.command
 net spindle-vel-fb-rpm      => pid.s.feedback
 net spindle-output        <=  pid.s.output
@@ -220,11 +237,20 @@
 net estop-out     <=  iocontrol.0.user-enable-out
 net estop-out     =>  iocontrol.0.emc-enable-in
 
-#  ---toolchange signals for custom tool changer---
+#  toolchanger
 
-net tool-number             <=  iocontrol.0.tool-prep-number
-net tool-change-request     <=  iocontrol.0.tool-change
-net tool-change-confirmed   =>  iocontrol.0.tool-changed
-net tool-prepare-request    <=  iocontrol.0.tool-prepare
-net tool-prepare-confirmed  =>  iocontrol.0.tool-prepared
+loadrt toolchanger
+addf toolchanger servo-thread
 
+net tool-change iocontrol.0.tool-change => toolchanger.toolchange
+net tool-changed iocontrol.0.tool-changed <= toolchanger.toolchanged
+net tool-number iocontrol.0.tool-prep-number => toolchanger.toolnumber
+net tool-oldnumber iocontrol.0.tool-number => toolchanger.currenttoolnumber
+
+net apos-cmd toolchanger.position-cmd => hm2_7i96.0.stepgen.02.position-cmd
+net toolchanger-stepgen-fb <= hm2_7i96.0.stepgen.02.position-fb => toolchanger.stepgen-pos-fb
+net tool-prepare-loopback iocontrol.0.tool-prepare => iocontrol.0.tool-prepared
+
+net estop-out <= iocontrol.0.user-enable-out
+#net estop-ext => iocontrol.0.emc-enable-in   stepgen.2.enable
+#net estop-ext <= iocontrol.0.emc-enable-in => toolchanger.enable => hm2_7i96.0.stepgen.02.enable
